File: packages/allstak-python/heavy_e2e_test.py
#!/usr/bin/env python3
"""Heavy E2E test for AllStak Python SDK — generates realistic telemetry load."""
import sys, os, time, uuid, traceback
sys.path.insert(0, os.path.join(os.path.dirname(__file__), 'src'))

import allstak

client = allstak.init(
    api_key="ask_live_4574x2yao33rtjbiuf2q873ltv6vpokb",
    host="http://localhost:8080",
    environment="e2e-heavy",
)
client.tracing.set_service("sdk-python-heavy-test")

trace_ids = []
span_count = 0
log_count = 0

# === A) TRACING: 5 successful + 2 error traces ===
for i in range(5):
    allstak.reset_trace()  # Force new trace per iteration
    root = allstak.start_span(f"HTTP GET /api/resource-{i}", description=f"Successful flow {i}")
    trace_ids.append(root.trace_id)

    child1 = allstak.start_span(f"DB SELECT resource_{i}", description="PostgreSQL query")
    time.sleep(0.02)
    child1.finish("ok")

    child2 = allstak.start_span(f"CACHE GET resource:{i}", description="Redis lookup")
    time.sleep(0.005)
    child2.finish("ok")

    if i % 2 == 0:
        child3 = allstak.start_span(f"API call external-svc/validate", description="External validation")
        time.sleep(0.015)
        child3.finish("ok")
        span_count += 1

    span_count += 3  # root + 2 children
    time.sleep(0.01)
    root.finish("ok")

# 2 error traces
for i in range(2):
    allstak.reset_trace()
    root = allstak.start_span(f"HTTP POST /api/orders (ERROR {i})", description=f"Error flow {i}")
    trace_ids.append(root.trace_id)
    child = allstak.start_span(f"DB INSERT orders (FAIL)", description="DB constraint violation")
    time.sleep(0.01)
    child.finish("error")
    span_count += 2
    root.finish("error")

print(f"Spans created: {span_count}")
print(f"Trace IDs: {trace_ids}")

# === B) LOGS: 15+ mixed levels ===
levels = [
    ("debug", "Initializing connection pool"),
    ("debug", "Cache warmup started"),
    ("info", "Server started on port 8080"),
    ("info", "Request received: GET /api/users"),
    ("info", "Query returned 42 rows"),
    ("info", "Response sent: 200 OK"),
    ("info", "Background job started: cleanup"),
    ("warn", "Slow query detected: 850ms"),
    ("warn", "Connection pool near capacity: 18/20"),
    ("warn", "Retry attempt 2/3 for external API"),
    ("error", "Failed to connect to Redis: timeout"),
    ("error", "Unhandled exception in request handler"),
    ("error", "Database deadlock detected"),
    ("fatal", "Out of memory: heap exhausted"),
    ("info", "Graceful shutdown initiated"),
    ("debug", "All connections closed"),
    ("info", "Server stopped"),
]

for level, msg in levels:
    tid = trace_ids[log_count % len(trace_ids)] if log_count < 10 else None
    getattr(allstak.log, level)(msg, service="sdk-python-heavy-test", environment="e2e-heavy",
                                  trace_id=tid if tid else None)
    log_count += 1

print(f"Logs sent: {log_count}")

# === C) HTTP REQUESTS: 10+ mixed ===
methods_paths = [
    ("GET", "/api/users", 200, 120), ("GET", "/api/users/1", 200, 45),
    ("POST", "/api/users", 201, 180), ("PUT", "/api/users/1", 200, 95),
    ("DELETE", "/api/users/99", 404, 30), ("GET", "/api/products", 200, 250),
    ("POST", "/api/orders", 201, 340), ("GET", "/api/orders/1", 200, 55),
    ("POST", "/api/payments", 500, 890), ("GET", "/api/health", 200, 5),
    ("PATCH", "/api/users/1", 200, 75), ("GET", "/api/reports", 200, 1200),
]

req_count = 0
for method, path, status, dur in methods_paths:
    tid = trace_ids[req_count % len(trace_ids)]
    allstak.http.record(
        trace_id=tid,
        direction="inbound",
        method=method,
        host="api.python-heavy.allstak.io",
        path=path,
        status_code=status,
        duration_ms=dur,
        request_size=100 + req_count * 10,
        response_size=500 + req_count * 50,
    )
    req_count += 1

print(f"HTTP requests sent: {req_count}")

# === D) ERRORS: 3 exceptions ===
errors_sent = 0
for exc_class, msg in [
    (ValueError, "Invalid user ID format"),
    (ConnectionError, "Database connection refused"),
    (RuntimeError, "Payment gateway timeout after 30s"),
]:
    try:
        raise exc_class(msg)
    except Exception as e:
        allstak.capture_exception(e)
        errors_sent += 1

print(f"Errors sent: {errors_sent}")

# === E) FAILURE BEHAVIOR: invalid key test ===
print("\n=== Testing failure behavior ===")
# Test invalid key via direct curl (SDK singleton prevents re-init)
import urllib.request, json
try:
    req = urllib.request.Request(
        "http://localhost:8080/ingest/v1/logs",
        data=json.dumps({"level":"info","message":"bad key test"}).encode(),
        headers={"Content-Type":"application/json","X-AllStak-Key":"invalid_key_xxx"},
        method="POST"
    )
    resp = urllib.request.urlopen(req)
    print(f"Bad key test: unexpected {resp.status}")
except urllib.error.HTTPError as e:
    print(f"Bad key test: correctly rejected with {e.code} ({e.reason})")
print("Bad key test: SDK did not crash (graceful failure)")

# Flush everything
allstak.flush()
time.sleep(2)
print("\n=== ALL DATA FLUSHED ===")
print(f"Total: {span_count} spans, {log_count} logs, {req_count} requests, {errors_sent} errors")
print(f"Trace IDs for verification: {trace_ids[:3]}...")

File: packages/allstak-js/heavy_e2e_test.mjs
import { AllStak } from './dist/node/index.mjs';

const client = AllStak.init({
  dsn: 'http://ask_live_4574x2yao33rtjbiuf2q873ltv6vpokb@localhost:8080',
  environment: 'e2e-heavy',
  tags: { service: 'sdk-js-heavy-test' },
});

const traceIds = [];
let spanCount = 0;

// A) TRACING: 5 successful + 2 error traces
for (let i = 0; i < 5; i++) {
  AllStak.resetTrace();
  const root = AllStak.startSpan(`HTTP GET /api/widgets-${i}`, { description: `JS success flow ${i}` });
  traceIds.push(root.traceId);

  const dbSpan = AllStak.startSpan(`DB SELECT widgets_${i}`, { description: 'MySQL query' });
  await sleep(20);
  dbSpan.finish('ok');

  const cacheSpan = AllStak.startSpan(`CACHE GET widgets:${i}`, { description: 'Memcached lookup' });
  await sleep(5);
  cacheSpan.finish('ok');

  if (i % 2 === 0) {
    const extSpan = AllStak.startSpan(`gRPC inventory.Check`, { description: 'External gRPC call' });
    await sleep(15);
    extSpan.finish('ok');
    spanCount++;
  }

  root.finish('ok');
  spanCount += 3;
}

for (let i = 0; i < 2; i++) {
  AllStak.resetTrace();
  const root = AllStak.startSpan(`HTTP DELETE /api/widgets/0 (FAIL ${i})`, { description: 'Error flow' });
  traceIds.push(root.traceId);
  const child = AllStak.startSpan(`DB DELETE widgets (CONSTRAINT)`, { description: 'FK violation' });
  await sleep(10);
  child.finish('error');
  root.finish('error');
  spanCount += 2;
}

console.log(`Spans: ${spanCount}`);
console.log(`Traces: ${traceIds.length} unique`);

// B) LOGS: 17 mixed
const logEntries = [
  ['debug', 'Express router initialized'],
  ['debug', 'Middleware chain: auth -> validate -> handle'],
  ['info', 'Server listening on :3000'],
  ['info', 'Incoming request: GET /api/widgets'],
  ['info', 'Cache miss, querying database'],
  ['info', 'Query returned 128 widgets'],
  ['info', 'Response sent: 200 OK, 45KB'],
  ['warn', 'Event loop lag: 150ms'],
  ['warn', 'Memory heap: 85% used'],
  ['warn', 'Unhandled promise rejection caught'],
  ['error', 'ECONNREFUSED: Redis connection failed'],
  ['error', 'Timeout: upstream service /api/inventory'],
  ['error', 'JSON parse error in request body'],
  ['info', 'Background worker: email queue processed'],
  ['debug', 'GC pause: 12ms'],
  ['info', 'WebSocket upgrade: /ws/notifications'],
  ['warn', 'Deprecated: Buffer() constructor'],
];

let logCount = 0;
for (const [level, msg] of logEntries) {
  const tid = logCount < 10 ? traceIds[logCount % traceIds.length] : undefined;
  AllStak.log[level](msg, { service: 'sdk-js-heavy-test', traceId: tid });
  logCount++;
}
console.log(`Logs: ${logCount}`);

// C) HTTP REQUESTS: 12 mixed
const requests = [
  ['GET', '/api/widgets', 200, 95], ['GET', '/api/widgets/1', 200, 30],
  ['POST', '/api/widgets', 201, 220], ['PUT', '/api/widgets/1', 200, 110],
  ['DELETE', '/api/widgets/5', 204, 40], ['GET', '/api/dashboard', 200, 350],
  ['POST', '/api/auth/login', 200, 180], ['POST', '/api/uploads', 413, 50],
  ['GET', '/api/search?q=test', 200, 800], ['GET', '/api/health', 200, 2],
  ['PATCH', '/api/widgets/1', 200, 75], ['GET', '/api/metrics', 500, 1500],
];

let reqCount = 0;
for (const [method, path, status, dur] of requests) {
  AllStak.captureRequest({
    traceId: traceIds[reqCount % traceIds.length],
    direction: 'inbound',
    method,
    host: 'api.js-heavy.allstak.io',
    path,
    statusCode: status,
    durationMs: dur,
    requestSize: 100,
    responseSize: 500,
    timestamp: Date.now(),
  });
  reqCount++;
}
console.log(`Requests: ${reqCount}`);

// D) ERRORS: 3
let errCount = 0;
for (const [name, msg] of [
  ['TypeError', 'Cannot read properties of undefined'],
  ['RangeError', 'Maximum call stack size exceeded'],
  ['SyntaxError', 'Unexpected token < in JSON at position 0'],
]) {
  const err = new Error(msg);
  err.name = name;
  AllStak.captureException(err);
  errCount++;
}
console.log(`Errors: ${errCount}`);

// E) FAILURE: bad key test
try {
  const resp = await fetch('http://localhost:8080/ingest/v1/logs', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', 'X-AllStak-Key': 'bad_key_xxx' },
    body: JSON.stringify({ level: 'info', message: 'bad key' }),
  });
  console.log(`Bad key test: HTTP ${resp.status} (expected 401)`);
} catch (e) {
  console.log(`Bad key test: ${e.message}`);
}

// Flush and exit — destroy triggers flush, then wait for network
AllStak.destroy();
await sleep(8000); // Wait longer for all batches to flush via network
console.log(`\n=== JS SDK DONE ===`);
console.log(`Trace IDs: ${traceIds.slice(0, 3).join(', ')}...`);

function sleep(ms) { return new Promise(r => setTimeout(r, ms)); }
